File: LINKED_LIST/JAVA/ReverseRec.java
import java.io.*;
import java.util.*;

class Node
{
   int data;
   Node next;

	   Node(int d)
	{
	   data=d;
	   next=null;

	}

}

 public class ReverseRec
{
        Node head;

	public void createNode(int new_data)
	{
           Node new_node= new Node(new_data);

	   new_node.next=head;
           head=new_node;
	}

   public void revRec()
   {
       revList(head);

   }
         public void revList(Node cur)
      {


        if(cur==null)
           return;

         if(cur.next==null)
         {
            this.head=cur;
             return;
         }

      revList(cur.next);

	cur.next.next=cur;
        cur.next=null;



  }


     public void printList()
     {
        while(head!=null){

 	System.out.print(head.data+ " ");
        head=head.next;
      }

     }

  public static void main(String args[])
{
   ReverseRec rev =new ReverseRec();

    rev.createNode(5);
    rev.createNode(8);
    rev.createNode(4);
    rev.createNode(7);
    rev.createNode(6);

    System.out.println(" before Reverse");
    rev.printList();

     System.out.println();
     rev.revRec();

     System.out.println("after reverse");

     rev.printList();


   }

}

File: LINKED_LIST/reverse_rec.c
#include<stdio.h>
#include<stdlib.h>
#include "create_node.h"


struct node *reverse_rec(struct node *head)
{
   struct node *cur=head;
   if(cur==NULL)
      return;
  if(cur->next==NULL)
  {
    head=cur;
    return;
  }

struct node *rev=reverse_rec(cur->next);

cur->next->next=cur;
cur->next=NULL;

return rev;
}

int main()
{
  struct node *start=NULL;
  create_node(&start,5);
  create_node(&start,8);
  create_node(&start,4);
  create_node(&start,9);
  create_node(&start,7);

  printf("list before reverse \n");
   printlist(start);
   struct node *temp=reverse_rec(start);
   printf("\n List after reverse \n");
   printlist(temp);

}
